javascript/ql/src/semmle/javascript/frameworks/Electron.qll

```diff
@@ -6,31 +6,444 @@ module Electron {
-  private abstract class BrowserObject extends DataFlow::NewNode {
+  abstract class NewBrowserObject extends BrowserObject, DataFlow::TrackedNode {
+    NewBrowserObject() {
+      this instanceof DataFlow::NewNode
+    }
+    
+    DataFlow::Node getWebPreferences() {
+      result = this.(DataFlow::NewNode).getOptionArgument(0, "webPreferences")
+    }
   }
   
   /**
    * A data flow node that creates a new `BrowserWindow`.
    */
-  class BrowserWindow extends BrowserObject {
+  class BrowserWindow extends NewBrowserObject {
     BrowserWindow() {
       this = DataFlow::moduleMember("electron", "BrowserWindow").getAnInstantiation()
     }
   }
   
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserWindow`
+   */
+  class TypedBrowserWindow extends BrowserObject {
+    TypedBrowserWindow() {
+      this.asExpr().getType().hasUnderlyingType("electron", "BrowserWindow")
+    }
+  }
+  
+  /**
+   * A data flow node with a TypeScript type indicating it is an Electron `BrowserView`
+   */
+  class TypedBrowserView extends BrowserObject {
+    TypedBrowserView() {
+      this.asExpr().getType().hasUnderlyingType("electron", "BrowserView")
+    }
+  }
   /**
    * A data flow node that creates a new `BrowserView`.
    */
-  class BrowserView extends BrowserObject {
+  class BrowserView extends NewBrowserObject {
     BrowserView() {
       this = DataFlow::moduleMember("electron", "BrowserView").getAnInstantiation()
     }
   }
-}
\ No newline at end of file
+  
+  /**
+   * A data flow node that is the `webContents` property of an Electron browser object
+   */
+  class WebContents extends DataFlow::TrackedNode {
+    WebContents() {
+      exists(BrowserObject bo |
+        bo = (this.(DataFlow::PropRead).getBase())
+      )
+      or
+      exists(NewBrowserObject nbo |
+        nbo.flowsTo(this.(DataFlow::PropRead).getBase())
+      )
+    }
+  }
+  
+  module IPC {
+    /**
+     * A data flow node that imports `electron.ipcMain`.
+     */
+    class Main extends DataFlow::SourceNode {
+      Main() {
+        this = DataFlow::moduleMember("electron", "ipcMain")
+      }
+    }
+    
+    /**
+     * A data flow node that is registered as an Electron IPC callback in the main process.
+     */
+    class MainCallback extends DataFlow::FunctionNode {
+      DataFlow::Node channel;
+      
+      MainCallback() {
+        exists(Main ipcMain, DataFlow::MethodCallNode mc |
+          mc = ipcMain.getAMemberCall("on") and
+          this.flowsTo(mc.getArgument(1)) and
+          channel = mc.getArgument(0)
+        )
+      }
+      
+      /**
+       * Gets the data flow node that is the channel the callback is listening on.
+       */
+      DataFlow::Node getChannel() {
+        result = channel
+      }
+      
+      /**
+       * Gets the string value that is the channel the callback is listening on.
+       */
+      string getChannelValue() {
+        result = channel.asExpr().getStringValue()
```

Comment:
That would lead to an overapproximation of flow if multiple strings flow to that argument; is that preferable to an underapproximation?

Reply:
That's up to you to decide. I'm thinking of cases where people use symbolic constants to represent channel names. While `mayHaveStringValue` isn't perfect for that due to the limited flow tracking it does, it would at least cover some simple cases.